guppylang/compiler/core.py

```diff
@@ -45,28 +21,49 @@ class DFContainer:
     current compilation state.
     """
 
+    graph: Hugr
     node: DFContainingNode
-    locals: CompiledLocals
-
-    def __getitem__(self, item: str) -> PortVariable:
-        return self.locals[item]
-
-    def __setitem__(self, key: str, value: PortVariable) -> None:
-        self.locals[key] = value
-
-    def __iter__(self) -> Iterator[PortVariable]:
-        return iter(self.locals.values())
-
-    def __contains__(self, item: str) -> bool:
-        return item in self.locals
+    locals: CompiledLocals = field(default_factory=dict)
+
+    def __getitem__(self, place: Place) -> OutPortV:
+        """Constructs a port for a local place in this DFG.
+
+        Note that this mutates the Hugr since we might need to pack or unpack some
+        tuples to obtain a port for places that involve struct fields.
+        """
+        # First check, if we already have a wire for this place
+        if place.id in self.locals:
+            return self.locals[place.id]
+        # Otherwise, our only hope is that it's a struct value that we can rebuild by
+        # packing the wires of its constituting fields
+        if not isinstance(place.ty, StructType):
+            raise InternalGuppyError(f"Couldn't obtain a port for `{place}`")
```

Comment:
What about the inverse? Could we auto-unpack structs when requesting a `FieldAccess`?

Reply:
Yes, that would be an alternative way of doing it: Instead of unpacking on assignments, we could unpack when requesting a field.

I actually implemented that version first, but the logic is more complicated. You need to be smart about looking upwards/downwards in the place hierarchy in the right order, to avoid doing an exponential search for all the possible ways to unpack/repack a value. Also, when you assign to a field you need to invalidate and unpack all parent places.

I think the benefit of the approach you suggested is that it leads to fewer unnessecary `MakeTuple`/`UnpackTuple` pairs.

But I wanted to do the most basic thing possible for this PR. Though I still have the other version around if you want to have a look if you prefer it?

Reply:
No need to block this now. Nicer hugrs is not a priority for guppy (though it would simplify tket2's logic :) ).